### PR TITLE
Collapse not shared NSG rules with multiple source ranges

### LINKED ISSUE #71482: Support collapse nsg rules with multiple source ranges
**What would you like to be added**:
When creating a service with annotation: `service.beta.kubernetes.io/load-balancer-source-ranges` containing multiple source ranges, the NSG should create a single rule with the source range collapsed. Currently, each source range listed in `service.beta.kubernetes.io/load-balancer-source-ranges` results in a new rule.

**Why is this needed**:
Support creation of more rules in the cloud provider

/sig azure
cc: @khenidak @weinong @feiskyer 

/kind feature

### PR DESCRIPTION
**Pull request by ritazh**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind flake

/kind feature

**What this PR does / why we need it**:
When creating a service with annotation: `service.beta.kubernetes.io/load-balancer-source-ranges` containing multiple source ranges and service.beta.kubernetes.io/azure-shared-securityrule: "false", the cloud provider will create a single rule with the source range collapsed in the NSG. This feature supports creation of more rules in the cloud provider.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #71482

**Special notes for your reviewer**:
Expected behavior for the following service annotations:
Not shared:
`service.beta.kubernetes.io/azure-shared-securityrule: "false"`
With source range provided:
`service.beta.kubernetes.io/load-balancer-source-ranges: "17.0.0.0/8,57.0.0.0/8"`
Internal:
`service.beta.kubernetes.io/azure-load-balancer-internal: "true"`

|Shared|With Source Range|Internal|Expected Behavior|
|---|---|---|---|
|n|y|n|UUID-PROTOCOL-PORT|
|n|n|n|UUID-PROTOCOL-PORT|
|n|y|y|UUID-PROTOCOL-PORT|
|n|n|y|no security rule will be created|
|y|y|n|shared-PROTOCOL-PORT-SOURCE-SourceRangePublicIP|
|y|n|n|shared-PROTOCOL-PORT-SOURCE-Internet|
|y|y|y|shared-PROTOCOL-PORT-SOURCE-SourceRangeInternalIP|
|y|n|y|no security rule will be created|

**Does this PR introduce a user-facing change?**:

```release-note
When creating a service with annotation: service.beta.kubernetes.io/load-balancer-source-ranges containing multiple source ranges and service.beta.kubernetes.io/azure-shared-securityrule: "false",  the NSG rules will be collapsed.
```

/sig azure
cc: @khenidak @weinong @feiskyer

### CODE REVIEW

**Comment by feiskyer** [comment]
> milestone v1.13

v1.13 is frozen now. Let's move this out of v1.13 milestone since this is not a critical bug fix.

File: pkg/cloudprovider/providers/azure/azure_standard.go [hunk 1]
```diff
@@ -236,11 +236,34 @@ func (az *Cloud) getSecurityRuleName(service *v1.Service, port v1.ServicePort, s
...
+	} else if len(sources) > 1 {
+		sourceAddressPrefix = ""
+		// this sort is used to ensure we get a predictable order to derive a name
+		// it is not an accurate sort
+		sort.Strings(sources)
+		sourceAddressPrefixes = sources
+	} else {
+		return sourceAddressPrefix, sourceAddressPrefixes, fmt.Errorf("sourceAddress should not be empty")
+	}
+	return sourceAddressPrefix, sourceAddressPrefixes, nil
```
Comment (feiskyer) [inline pkg/cloudprovider/providers/azure/azure_standard.go]: could we merge these two together into `sourceAddressPrefixes`? This could make it easier to understand.

File: pkg/cloudprovider/providers/azure/azure_standard.go [hunk 2]
```diff
@@ -236,11 +236,34 @@ func (az *Cloud) getSecurityRuleName(service *v1.Service, port v1.ServicePort, s
 		safePrefix := strings.Replace(sourceAddrPrefix, "/", "_", -1)
 		return fmt.Sprintf("shared-%s-%d-%s", port.Protocol, port.Port, safePrefix)
 	}
+	if sourceAddrPrefix == "" {
```
Comment (feiskyer) [inline pkg/cloudprovider/providers/azure/azure_standard.go]: sourceAddrPrefix is default to "Internet" by default, so why is this required?

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go [hunk 3]
```diff
@@ -1069,33 +1080,40 @@ func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service,
...
-					updatedRules = append(updatedRules[:sharedIndex], updatedRules[sharedIndex+1:]...)
-				} else {
-					newDestinations := append(existingPrefixes[:addressIndex], existingPrefixes[addressIndex+1:]...)
-					sharedRule.DestinationAddressPrefixes = &newDestinations
-					updatedRules[sharedIndex] = sharedRule
-				}
-				dirtySg = true
+			sharedRuleName := ""
+			if len(expectedSourceAddressPrefixes) > 1 {
+				sharedRuleName = az.getSecurityRuleName(service, port, strings.Join(expectedSourceAddressPrefixes, "-"))
```
Comment (feiskyer) [inline pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: `strings.Join(expectedSourceAddressPrefixes, "-")` would make the rule name different if sourceAddressPrefixes are updated (e.g. delete one by kubectl edit service). So the following `findSecurityRuleByName()` would always find nothing in this case?

File: pkg/cloudprovider/providers/azure/azure_standard.go [hunk 4]
```diff
@@ -236,11 +236,34 @@ func (az *Cloud) getSecurityRuleName(service *v1.Service, port v1.ServicePort, s
...
+		rulePrefix := az.getRulePrefix(service)
+		return fmt.Sprintf("%s-%s-%d", rulePrefix, port.Protocol, port.Port)
+	}
+	// ensure backward compatibility
 	safePrefix := strings.Replace(sourceAddrPrefix, "/", "_", -1)
 	rulePrefix := az.getRulePrefix(service)
 	return fmt.Sprintf("%s-%s-%d-%s", rulePrefix, port.Protocol, port.Port, safePrefix)
 }
 
+func (az *Cloud) getSourceAddressPrefix(sources []string) (string, []string, error) {
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure_standard.go]: `getSourceAddressPrefix` does not use az member field, it's better remove `(az *Cloud)` and also add a unit test for this func

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go [hunk 5]
```diff
@@ -1069,33 +1080,40 @@ func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service,
...
+				return nil, fmt.Errorf("Expected to find shared rule %s for service %s being deleted, but did not", sharedRuleName, service.Name)
+			}
+			if sharedRule.DestinationAddressPrefixes == nil {
+				klog.V(2).Infof("Expected to have array of destinations in shared rule for service %s being deleted, but did not", service.Name)
+				return nil, fmt.Errorf("Expected to have array of destinations in shared rule for service %s being deleted, but did not", service.Name)
 			}
+			existingPrefixes := *sharedRule.DestinationAddressPrefixes
+			addressIndex, found := findIndex(existingPrefixes, destinationIPAddress)
+			if !found {
+				klog.V(2).Infof("Expected to find destination address %s in shared rule %s for service %s being deleted, but did not", destinationIPAddress, sharedRuleName, service.Name)
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: Errorf

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go [hunk 6]
```diff
@@ -1069,33 +1080,40 @@ func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service,
...
-				dirtySg = true
+			sharedRuleName := ""
+			if len(expectedSourceAddressPrefixes) > 1 {
+				sharedRuleName = az.getSecurityRuleName(service, port, strings.Join(expectedSourceAddressPrefixes, "-"))
+			} else {
+				sharedRuleName = az.getSecurityRuleName(service, port, expectedSourceAddressPrefix)
+			}
+			sharedIndex, sharedRule, sharedRuleFound := findSecurityRuleByName(updatedRules, sharedRuleName)
+			if !sharedRuleFound {
+				klog.V(2).Infof("Expected to find shared rule %s for service %s being deleted, but did not", sharedRuleName, service.Name)
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: Errorf

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go [hunk 7]
```diff
@@ -1069,33 +1080,40 @@ func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service,
...
+				return nil, fmt.Errorf("Expected to find shared rule %s for service %s being deleted, but did not", sharedRuleName, service.Name)
+			}
+			if sharedRule.DestinationAddressPrefixes == nil {
+				klog.V(2).Infof("Expected to have array of destinations in shared rule for service %s being deleted, but did not", service.Name)
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: Errorf

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go [hunk 8]
```diff
@@ -1006,35 +1006,46 @@ func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service,
...
-						Access:                   network.SecurityRuleAccessAllow,
-						Direction:                network.SecurityRuleDirectionInbound,
-					},
+			expectedSourceAddressPrefix, expectedSourceAddressPrefixes, err := az.getSourceAddressPrefix(sourceAddressPrefixes)
+			if err != nil {
+				return nil, err
+			}
+			var securityRuleName string
+			if useSharedSecurityRule(service) {
+				klog.V(2).Infof("Reconcile security group for service %s. When service is shared, the rule name includes sourceAddressPrefixes", service.Name)
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: is it better put this `When service is shared, the rule name includes sourceAddressPrefixes` into code comment?

**Review by andyzhangx — COMMENTED** [review]
you could run `hack/verify-gofmt.sh` to check code format error before sending out PR.

File: pkg/cloudprovider/providers/azure/azure_standard.go (hunk 2 above)
Comment (ritazh) [reply pkg/cloudprovider/providers/azure/azure_standard.go]: https://github.com/kubernetes/kubernetes/blob/ec4b3a231cfa19b8cfc5c5b8fbf7fac048914f87/pkg/cloudprovider/providers/azure/azure_loadbalancer.go#L1021-L1030

The logic for security rule name is changed such that if it's a shared rule, we append `Internet` or the sourceAddressPrefix, for example:`shared-PROTOCOL-PORT-Internet` and `shared-PROTOCOL-PORT-SOURCE-SourceRangePublicIP` respectively. If it's a not shared rule, then we do not append anything, for example: `UUID-PROTOCOL-PORT` since UUID is unique and we don't want to create a different rule name for each sourceAddressPrefix.

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 3 above)
Comment (ritazh) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: Yes unfortunately that's the behavior today. Since the rule name includes sourceAddressPrefixes, if you edit a service's source range, the cloud provider will think it's a new rule and create a new rule while the old rule with the old sourceAddressPrefixes in its name stays around forever.

With this PR, this behavior is only the case for shared rules. For rules that are not shared,  editing a service's sourceAddressPrefixes will not affect the rule name so it will still be deleted even if user's changed the service's sourceAddressPrefixes.

Should updating `load-balancer-source-ranges` as part of `kubectl edit service` be a supported scenario?

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 5 above)
Comment (ritazh) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: https://github.com/kubernetes/kubernetes/blob/f13ae7cf1121133f75db6bffff9f311dabba1f75/pkg/cloudprovider/providers/azure/azure_loadbalancer.go#L1096-L1097

Errorf is right below this line

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 6 above)
Comment (ritazh) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: https://github.com/kubernetes/kubernetes/blob/f13ae7cf1121133f75db6bffff9f311dabba1f75/pkg/cloudprovider/providers/azure/azure_loadbalancer.go#L1096-L1097

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 3 above)
Comment (feiskyer) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: yep, `loadBalancerSourceRanges` is part of service spec, it may be changed later after creation.

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 5 above)
Comment (andyzhangx) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: if Error is returned and ultimately shows in the log or events, then we don't need to print earlier, otherwise it will print twice.

File: pkg/cloudprovider/providers/azure/azure_standard.go (hunk 2 above)
Comment (feiskyer) [reply pkg/cloudprovider/providers/azure/azure_standard.go]: could we keep it same as before? so that if the cluster is upgraded from old versions, old rules could still be recognized.

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 3 above)
Comment (ritazh) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: > yep, loadBalancerSourceRanges is part of service spec, it may be changed later after creation.

But the existing behavior when loadBalancerSourceRanges is updated is that a new rule gets created and the old rule just lives on forever since the rule name can no longer be matched.

File: pkg/cloudprovider/providers/azure/azure_standard.go (hunk 2 above)
Comment (ritazh) [reply pkg/cloudprovider/providers/azure/azure_standard.go]: If we keep the rule name the same as before, then we cannot collapse the rules into one as the name includes the source range as part of it. Do you have any suggestions?

File: pkg/cloudprovider/providers/azure/azure_loadbalancer.go (hunk 3 above)
Comment (feiskyer) [reply pkg/cloudprovider/providers/azure/azure_loadbalancer.go]: oops, that should be a bug I didn't notice. We should fix that

**Comment by ritazh** [comment]
@feiskyer I have updated this PR to support updating `service.beta.kubernetes.io/load-balancer-source-ranges` when `service.beta.kubernetes.io/azure-shared-securityrule: "false"`. As well as backward compatibility to support existing non-shared rules `UUID-PROTOCOL-PORT-SOURCEIP`.  

https://github.com/kubernetes/kubernetes/blob/c2c9f7e369c7cb6478b9b09a42f6fc74b2dd944d/pkg/cloudprovider/providers/azure/azure_loadbalancer.go#L1140-L1145

When `service.beta.kubernetes.io/azure-shared-securityrule: "true"`, the behavior is the same as today.

**Comment by feiskyer** [comment]
@ritazh Cool, thanks for the update.  The change LGTM.

/kind bug
/priority important-soon
/retest

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Comment by andyzhangx** [comment]
@ritazh will you also update the PR description? looks like it's not identical to your initial description.
And also squash all commits, thanks. I will then LGTM, thanks.
/lgtm cancel

**Review by feiskyer — COMMENTED** [review]
/lgtm

**Review by andyzhangx — APPROVED** [review]
/approve
